core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java

```diff
@@ -619,6 +620,7 @@ public RelOptPredicateList inferPredicates(
 
       final RexBuilder rexBuilder = joinRel.getCluster().getRexBuilder();
       switch (joinType) {
+      case SEMI:
```

Comment:
L627~629 should be pulled up here. SEMI can separate with INNER.

Reply:
I agree, separate SEMI and INNER logic, also remove the isSemiJoin flag(useless now).